datapoint/datapoint.go

```diff
@@ -48,6 +50,40 @@ type Datapoint struct {
 	Timestamp time.Time `json:"timestamp"`
 }
 
+type jsonDatapoint struct {
+	Metric     string            `json:"metric"`
+	Dimensions map[string]string `json:"dimensions"`
+	Value      interface{}       `json:"value"`
+	MetricType MetricType        `json:"metric_type"`
+	Timestamp  time.Time         `json:"timestamp"`
+}
+
+// UnmarshalJSON decodes JSON into a datapoint, creating the correct Value interface types for the
+// type of JSON value that was encoded
+func (dp *Datapoint) UnmarshalJSON(b []byte) error {
+	var m jsonDatapoint
+	dec := json.NewDecoder(bytes.NewBuffer(b))
+	dec.UseNumber()
+	if err := dec.Decode(&m); err != nil {
+		return err
+	}
+	switch t := m.Value.(type) {
+	case string:
+		dp.Value = NewStringValue(t)
+	case json.Number:
+		if num, e := t.Int64(); e == nil {
+			dp.Value = NewIntValue(num)
+		} else if num, e := t.Float64(); e == nil {
+			dp.Value = NewFloatValue(num)
```

Comment:
Potentially we could have no value?  Is that ok? If not possible, this can just be an else right?